docs/specification.rst

```diff
@@ -63,606 +61,407 @@ For example,
-- memoryview
-- range
-- set
-- str
-- tuple
-- type
-
-There are concrete types that can't be subclassed, they are
- - a class marked with ``typing.final`` `[doc] <https://docs.python.org/3/library/typing.html#typing.final>`_
- - ``typing.Never`` and ``typing.NoReturn`` also called `bottom type <https://en.wikipedia.org/wiki/Bottom_type>`_
- - ``None``
-
-If such a type is used within an intersection this intersection shall evaluate to ``Never``.
-
-The reasoning behind this is that these types can't be subtyped and shouldn't be dynamically
-extended.
-Doing this early prevents issues during subtyping or assignments checks.
 
-::
+``Intersection`` does not forbid any incompatibility of type parameters
+(Neither statically or at runtime).
 
-    from typing import TypeVar, reveal_type
 
-    T = TypeVar("T")
+Runtime specification behavior
+------------------------------
 
-    class Enhanced:
-        is_great: bool
+At runtime, ``Intersection[*Ts]`` and ``Intersection[TypeOne, TypeTwo]`` each create an
+object which can be introspected consistent with the methods provided for type
+introspection in ``typing`` such as, but not limited to, ``get_origin``.
 
 
-    def enhance(cls: type[T]) -> type[T & Enhanced]:
-        class New(cls, Enhanced):
-            ...
 
-        return New
+Rationale
+=========
 
-    reveal_type(enhance(str))  # okay
-    reveal_type(enhance(None))  # raises a TypeError on runtime, should be flagged by TypeCheckers
+The intersection discussion was long and complex, with many edge cases explored
+to determine the feasibility of an intersection. In this section I will
+summarize why certain design decisions were made.
 
-It is important to note that once a type checker evaluated anything to ``Never`` within an
-intersection it can stop further evaluations an return ``Never``.
-This way a lot of edge cases by mixin types that can't be mixed are handled easily.
+Inheritance
+-----------
 
-Handling Callables
-------------------
-Every Callable within an intersection shall be treated like a ``def __call__()`` Protocol.
+This was never very controversial - the original idea was to have intersection
+reflect the way that union works. For ``Union[A,B]``, it follows that:
 
 ::
 
-    from typing import Protocol, Callable
+    x: Union[A,B]
+    isinstance(x, A) or isinstance(x, B) # Always true
 
-    MyCallable = Callable[[str, int], float]
+So for Intersection it also follows that:
 
-    class CallProto:
-        def __call__(a: str, b: int) -> float: ...
+::
 
-    # Type Checker should perform the following conversion
-    # T & MyCallable => T & CallProto
+    x: Intersection[A,B]
+    isinstance(x, A) and isinstance(x,B) # Always true
 
-This way the ``overload`` mechanism described below can be used.
+The only discrepancy here is that would mean it must be a class that inherits from
+both A and B. In inheritance the order matters, so suddenly this means that the order
+matters for intersection as well.
 
+Ordering
```

Comment:
Hmm this is a new idea - personally I think allowing type checkers to diverge on this would lead to quite wildly different behaviour, but it might be worth exploring. From comments I've seen before on similar ideas I think people would probably just consider this "kicking the can down the road" which is a fair enough comment.

> Is it worth discussing also that Union is unordered, and thus by introducing ordering for Intersection we require users to be aware of this subtlety?

The ordering is explored quite a bit in the how to teach section in order to resolve this - do you think this section makes it clearer?